scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py

```diff
@@ -368,11 +368,23 @@ def lobpcg(
         else:
             eigvals = (0, sizeX - 1)
 
-        A_dense = A(np.eye(n, dtype=A.dtype))
-        B_dense = None if B is None else B(np.eye(n, dtype=B.dtype))
+        if isinstance(A, LinearOperator):
+            A = A(np.eye(n, dtype=A.dtype))
+        elif isspmatrix(A):
+            A = A.toarray()
+        else:
+            A = np.asarray(A)
+
+        if B is not None:
+            if isinstance(B, LinearOperator):
+                B = B(np.eye(n, dtype=B.dtype))
+            elif isspmatrix(B):
+                B = B.toarray()
+            else:
+                B = np.asarray(B)
```

Comment:
Not sure if it is practical to add tests for these cases, but a few lines don't get hit in the new code according to the codecov report:

![image](https://user-images.githubusercontent.com/7903078/147694396-3b69cc38-358c-4adf-9934-3e9022fdc62d.png)

Reply:
@tylerjereddy - Thanks for reviewing! I'll have a look. It's possible that these cases have never historically appeared in unit tests. Probably might not be too difficult to add, and it's probably worth doing.

Reply:
OK, I have added tests for A, B, and M in the form of dense array, sparse matrix, and LinearOperator both for the special situation ((n - sizeY) < (5 * sizeX)) and not.
The failed tests come from stats, i.e. unrelated to this PR.